system/firewalld.py

```diff
@@ -401,9 +408,20 @@ def main():
         modification_count += 1
     if masquerade != None:
         modification_count += 1
+    if default_zone:
+        modification_count += 1
 
     if modification_count > 1:
-        module.fail_json(msg='can only operate on port, service, rich_rule or interface at once')
+        module.fail_json(msg='can only operate on port, service, rich_rule, '
+                         'interface or default_zone at once')
+
+    if (default_zone is None) == (desired_state is None):
+        module.fail_json(msg='mandatory options are either (default_zone, '
+                         'permanent, immediate) or (state, permanent)')
+
+    if default_zone and (not permanent or not immediate):
+        module.fail_json(msg='default_zone can be used only with options '
+                         'permanent and immediate set to True')
 
```

Comment:
I'd like to see a check here that makes sure that `desired_state` has a non-`None` value for all options other than `default_zone` so that for all other options that aren't using `desired_state` the functionality is the same (i.e. - it would fail in the event that `state` parameter is omitted)

Maybe something along the lines of:
```python
if (default_zone is None) and (desired_state is None):
    module.fail_json(msg='"state" is a mandatory option when not performing an operating on "default_zone"')
```